### DISCUSSION TITLE
Protege as tool for taxonomy modeling.

### DISCUSSION THREAD

**Discussion by DavidJurajdaNXP** [Open-CMSIS-Pack/Open-CMSIS-Pack-Spec#251] [Ideas]
Software component classification structure is affected by end user perspective. Since that, multiple taxonomies might be relevant.

I am proposing to use https://protege.stanford.edu/ tool for cmsis pack taxonomy modeling.

Related tickets:
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/135 Restrict Cclass values to pre-defined list
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/22 Taxonomic rank shift
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/94 Separation of component identification and classification
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/23 Component Discovery Gaps 
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/114 API, Components, and Interfaces - better definition required
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/157 Yet another doubt about the concept of component identity...

Input data extracted from *.pdsc files:
[cbundle.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485362/cbundle.txt)
[cclass.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485363/cclass.txt)
[cgroup.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485364/cgroup.txt)
[csub.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485365/csub.txt)
[cvariant.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485366/cvariant.txt)
[explicit_api_tree.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485368/explicit_api_tree.txt)
[explicit_api_tree_cmsis_only.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485369/explicit_api_tree_cmsis_only.txt)
[implicit_api_tree.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12485370/implicit_api_tree.txt)

Protege project files:
[embedded.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12486168/embedded.txt)
[explicit_cmsis_api.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12486169/explicit_cmsis_api.txt)
[explicit_cmsis_api_arm.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12486171/explicit_cmsis_api_arm.txt)
[implicit_cmsis_api.txt](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/files/12486172/implicit_cmsis_api.txt)

screenshot:
![image](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/assets/16876353/546f71e6-619f-4ddd-a599-28b9603adde5)

https://www.nist.gov/publications/feature-technology-and-ontology-embedded-system-design-and-development
https://en.wikipedia.org/wiki/Feature_model

**Comment by DavidJurajdaNXP** [comment]
**What is benefit of ontology over schema (xsd, database ER)?**

It is very similar. Ontology is more focused on meaning. See:  
- https://pdfs.semanticscholar.org/b44f/a4592b69183c1965d0075dea1a3bc58dfbfe.pdf
- https://www.w3.org/wiki/SchemaVsOntology
- http://www.dl.edi-info.ir/Ontology%20and%20database%20schema,%20What%20is%20the%20difference.pdf

I think we should explore how to classify single entity in multiple taxonomies. 
Example of alternative taxonomy might be [Feature based design ](https://www.nist.gov/publications/feature-technology-and-ontology-embedded-system-design-and-development).

**Comment by DavidJurajdaNXP** [comment]
Discussion about taxonomy/ontology/schema can be observed also from tools perspective. What is your favorite editor?

**Comment by DavidJurajdaNXP** [comment]
Let's take a look at Eclipse XML schema editor:
![image](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/assets/16876353/95d4b387-47dc-4cc9-8860-e209bfef5ccc)
we can see list of types on the left side of view, but component and bundle are not there, because this types are nested in package description. 

If we compare this view with Protege perspective, we now the component is significant entity in both views, but in case of xml schema it not immediately visible in graphical view.

The difference might be mapped to difference between [Conceptual data model](https://en.wikipedia.org/wiki/Conceptual_data_model) and [Logical data model](https://en.wikipedia.org/wiki/Logical_data_model) as described in https://en.wikipedia.org/wiki/Data_model

**Comment by iomint** [comment]
Hi David,
I am not sure to catch what you encompass in "taxonomy modeling". Are you looking at "Protege" as a taxonomy editor/documentation tool  or more ? I have to say that I did not had a lot of time to explore the capabilities of it so if it is more it will be great if you can detailed further the usage you have in mind otherwise but it looks to me a bit overkill in our context . 
Eric

**Reply by DavidJurajdaNXP** [reply]
Hi @iomint,

I would define "taxonomy modeling" as formal description of embedded systems domain.  I would like to achieve it via visualization (collapsible hierarchy, graph), formalization (OWL/RDF) and annotation (comments, metadata).

since we need to formalize open cmsis pack taxonomy somehow, it seems logical to me to select already existing tool.
Protege provides multiple standardized formats (owl,rdf...), graphical interface and reasoners which provide space to expand in the future.

I am aiming also for future applications like knowledge graphs and graph neural networks (https://pyg.org/).
We should design our data in way which would directly support such scenarios. In this context plain classes should be enriched by additional rules.

OWL format is supported also by libraries like https://owlready2.readthedocs.io/en/latest/index.html#

I am open to any other solutions like: CMSIS.Taxonomy.pdsc proposed in: https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/252

We would need to compare definition formats and visual appearance in tools to select right approach.
Regarding to overkill, I can imagine we might start with something simple like plain text, but I am expecting we quickly reach its limits.

I am looking forward for any proposals to compare.

**Comment by DavidJurajdaNXP** [comment]
Here are python package index classifiers and topics:
https://pypi.org/search/
https://pypi.org/classifiers/

**Reply by KeilChris** [reply]
I don't see how these classifiers differ from what we have today apart from having more "levels" than the three we have now.

**Reply by DavidJurajdaNXP** [reply]
Hi @KeilChris, 

python classifiers above are mentioned as example of classification in different framework.

Key differences are:
- more levels (as you mentioned)
- multiple classifiers can be assigned to single entity
- classifiers are not entity identifiers

Consider following issue:
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/22 Taxonomic rank shift

In current structure (3 levels), it can be solved by shifting of "CMSIS" category to higher level (framework?) or by defining different classification schemas to different entity types (API, sw component).

Coupling of identification and classification together introduce obstacle for future taxonomy development, when existing entity would fit better to newly created category.
